tds-prize-tool/src/rewards_earned.rs

```diff
@@ -0,0 +1,289 @@
+//! Calculates the winners of the "Most Rewards Earned" category in Tour de Sol by summing the
+//! balances of all stake and vote accounts attributed to a particular validator.
+//!
+//! The top 3 validators will receive the top prizes and validators will be awarded additional
+//! prizes if they place into the following buckets:
+//!
+//! `hi` - Top 25%
+//! `md` - Top 25-50%
+//! `lo` - Top 50-90%
+
+use crate::prize::{self, Winner, Winners};
+use solana_runtime::bank::Bank;
+use solana_sdk::account::Account;
+use solana_sdk::pubkey::Pubkey;
+use solana_stake_api::stake_state::StakeState;
+use solana_vote_api::vote_state::VoteState;
+use std::cmp::{max, min};
+use std::collections::HashMap;
+
+pub const STARTING_BALANCE_LAMPORTS: i64 = 0; // TODO
+
+const HIGH_BUCKET: &str = "Top 25% Bucket";
+const MID_BUCKET: &str = "Top 25-50% Bucket";
+const LOW_BUCKET: &str = "Top 50-90% Bucket";
+
+fn voter_stake_rewards(stake_accounts: HashMap<Pubkey, Account>) -> HashMap<Pubkey, u64> {
+    let mut voter_stake_sum: HashMap<Pubkey, u64> = HashMap::new();
+    for (_key, account) in stake_accounts {
+        if let Some(StakeState::Stake(_authorized, _lockup, stake)) = StakeState::from(&account) {
+            let stake_sum = voter_stake_sum
+                .get(&stake.voter_pubkey)
+                .cloned()
+                .unwrap_or_default();
+            voter_stake_sum.insert(stake.voter_pubkey, stake_sum + account.lamports);
+        }
+    }
+    voter_stake_sum
+}
+
+fn validator_results(validator_reward_map: HashMap<Pubkey, u64>) -> Vec<(Pubkey, i64)> {
+    let mut validator_rewards: Vec<(Pubkey, u64)> = validator_reward_map
+        .iter()
+        .map(|(key, balance)| (*key, *balance))
+        .collect();
+
+    // Sort descending and calculate results
+    validator_rewards.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap());
+    validator_rewards
+        .into_iter()
+        .map(|(key, earned)| (key, (earned as i64) - STARTING_BALANCE_LAMPORTS))
+        .collect()
+}
+
+fn validator_rewards(
+    mut voter_stake_rewards: HashMap<Pubkey, u64>,
+    vote_accounts: HashMap<Pubkey, (u64, Account)>,
+) -> HashMap<Pubkey, u64> {
+    // Sum validator earned reward totals (stake rewards + commission)
+    let mut validator_reward_map: HashMap<Pubkey, u64> = HashMap::new();
+    for (voter_key, (_stake, account)) in vote_accounts {
+        if let Some(vote_state) = VoteState::from(&account) {
+            let voter_commission = account.lamports;
+            let voter_stake_reward = voter_stake_rewards.remove(&voter_key).unwrap_or_default();
+
+            // TODO: check if it's possible to have multiple vote accounts in working bank
```

Comment:
it is ;)

Reply:
🙊

Reply:
it's not a reasonable setup, as our implementation only supports voting to one account at a time, but it's possible to have stake dedicated to a single node via 2 accounts